### DISCUSSION TITLE
using ConformingDelaunayTriangulationBuilder, remove triangles outside of the bounding polygon

### DISCUSSION THREAD

**Discussion by southernsun** [NetTopologySuite/NetTopologySuite#772] [Q&A] [answered]
How do you effectively remove triangles from a ConformingDelaunayTriangulationBuilder build that are outside of a defined bounding polygon? As in the example below, the outside polygons skew/messes up the contour lines.

In the example below, the red line is the bounding polygon. And I want to remove the triangles that are marked yellow, as they are outside the bounding polygon.

![image](https://github.com/user-attachments/assets/abb7a898-a9d0-418a-9928-9c0910ae7bb3)

I've seen some references to using the Constraint property or GetEdges(), but that didn't seem to solve the issue for me. I'm thinking i have to implement a routine that evaluates all the triangles against the boundary. If so, what is the most effective way and with the best performance of doing that? Would you recommend using an STRTree?

Any insights or experience from the community is appreciated. Thanks!

**Comment by southernsun** [comment]
Removing triangles that overlap causes another issues, as the data point is just outside the bounding polygon. Discarding those polygons create "holes" in my triangles dataset.

![image](https://github.com/user-attachments/assets/e03d397e-0d4f-46d6-aa63-2428fd932b76)

So do i need to re-adjust those triangles?

I assume using Constraints property on the builder is the better option, but that will throw a LocateFailureException:
NetTopologySuite.Triangulate.QuadEdge.LocateFailureException: 'Locate failed to converge (at edge: LINESTRING( -103.51661423108342 43.86764509014661, -103.516618 43.869728)).  Possible causes include invalid Subdivision topology or very close sites'

**Comment by southernsun** [comment]
I actually resolved the LocateFailureException when using the Constraint property by reducing the ridicules precision on the lat/long coordinates to 5. The input file I'm using has up to 12. But i still have triangles on the outside of the bounding area.

**Reply by FObermaier** [reply, marked as answer]
You need to create an `IndexedPointInAreaLocator` using the Polygon that is being triangulated. Use that to test if the centroid of the triangle is inside the area or not.

**Reply by southernsun** [reply]
thanks, i will have a look.

**Reply by southernsun** [reply]
Filtering on centroid using IndexedPointInAreaLocator cleared-up most issues, but i have this strange bounding box polygon which is probably incorrect but i want to handle this accordingly:

![image](https://github.com/user-attachments/assets/4aa9b8b3-9db6-4583-ab6b-f9dff73f8a5a)

Would it be wise to search for all triangles that are intersected by the bounding box polygon and use the intersecting line to split the triangles into two polygons (recalculating the associated value with the vertices) and discard the polygon that is outside the bounding box polygon? And how would you convert the remaining polygon into a valid triangle, for example the orange outline would be a result of the above described routine:

![image](https://github.com/user-attachments/assets/01b35039-472f-43fe-945f-d36877950b1c)

Any insights on how you would approach this are more than welcome. Thank you for your time.

**Reply by FObermaier** [reply]
Did you set add the boundary as a constraint to the triangulation builder?

**Reply by southernsun** [reply]
That will fix the issue, but several of the Tri objects generated on the edge won't have associated M value recalculated on the vertices.

![image](https://github.com/user-attachments/assets/075eaddd-9c75-40b2-992f-a1c8f9c65954)

What would be the most effective way to calculate a M value for these coordinates based on the original grid? Find the closest two points of the input grid and interpolate the M value between them for the vertices of the Tri that don't have an associated M value?

**Reply by southernsun** [reply]
I'm considering doing Inverse Distance Weighting, although my current implementation is still a bit slow.
